📝 Add Intermediate Event Docu (#23)
atlas-engine/ContributionGuidelines: docs/getting_started/bpmn_elements/intermediate_event.md

# Intermediate Event

An Intermediate Events are, unlike Boundary Events, not attached to an activity.

Basically there are two different kinds of Intermediate Events: Intermediate Catch Events and Intermediate Throw Events.

## Intermediate Catch Event

Intermediate Catch Events are used if the process needs to wait for a specific trigger to happen.

## Intermediate Throw Event

> Intermediate Throw Events on the other hand are used to trigger those Intermediate Catch Events.

Author: Stimmt natürlich

Both the Intermediate Message Catch Event and the Intermediate Message Throw Event must be configured with the same Message Event in order to work.

#### Example

![Intermediate Message Event Example](images/intermediate_event/intermediate_message_event_example.svg)

In this example the `Do some stuff` Task and the Intermediate Message Catch Event are triggered as soon as the Process is started.
The Intermediate Message Catch Event waits until it recevies its specified Message Event.
When the `Do some stuff` Task has finished the Intermediate Message Throw Event will trigger the Message Event, that is also configured in the Intermediate Message Catch Event.
The Intermediate Message Catch Event whereupon finish which executes the `Do some more stuff` Task.

### Intermediate Signal Event

![Intermediate Signal Throw Event](images/intermediate_event/intermediate_signal_throw_event.svg)
![Intermediate Signal Catch Event](images/intermediate_event/intermediate_signal_catch_event.svg)

The Intermediate Signal Event is used to pause a process execution until a specific Signal was received.

#### Configuration

> To configure a Intermediate Signal Event a Signal Event must be defined:

Reviewer: ```suggestion
To use an Intermediate Signal Event, a Signal must be attached to it:
```

<img width="350px" alt="Intermediate Signal Event PropertyPanel" src="images/intermediate_event/intermediate_signal_event_property_panel.png">

> Both the Intermediate Signal Catch Event and the Intermediate Signal Throw Event must be configured with the same Signal Event in order to work.

Reviewer: ```suggestion
Both the Intermediate Signal Catch Event and the Intermediate Signal Throw Event must use a signal with the same name in order to be able to communicate with each other.
```

#### Example

![Intermediate Signal Event Example](images/intermediate_event/intermediate_signal_event_example.svg)

In this example the `Do some stuff` Task and the two Intermediate Signal Catch Event are triggered as soon as the Process is started.
The Intermediate Signal Catch Events wait until they recevies their specified Signal Event.
When the `Do some stuff` Task has finished the Intermediate Signal Throw Event will trigger the Signal Event, that is also configured in the two Intermediate Signal Catch Event.

> The Intermediate Signal Catch Events whereupon finish which executes the `Do some more stuff` Task and the `Do some other stuff` Task.

Reviewer: Hier das gleiche wie beim Message Event Sample.

> ### Intermediate Link Event

Reviewer: Ich würde das Link Event als letztes beschreiben, da es das am wenigsten relevante ist.

![Intermediate Link Throw Event](images/intermediate_event/intermediate_link_throw_event.svg)
![Intermediate Link Catch Event](images/intermediate_event/intermediate_link_catch_event.svg)

The Intermediate Link Event is used to connect two sections of a process.

Only one Intermediate Link Catch Event can exist for each link name.

> Also the Link Event must be on the same process level, for that reason it is not possible to send a Link Event between a Subprocess and its parent process.

Reviewer: ```suggestion
Also the Link Event must be on the same process level.
For this reason, it is not possible to send a Link Event between a Subprocess and its parent process.
It is also not possible to use a Link Throw Event to access the Link Catch Event of another Process Model.
```

> #### Configuration

Reviewer: ```suggestion
#### How to use it
```

>  The only information that is needed to configure an Intermediate Link Event is the Link Name
>  The Intermediate Link Catch Event and the Intermediate Link Throw Event both must use the exact same Link Name in order to work correctly.

Reviewer: ```suggestion
The only information that is needed to configure an Intermediate Link Event is the Link Name.
Both the Intermediate Link Catch Event and the Intermediate Link Throw Event must use the exact same Link Name in order to work correctly.
```

<img width="350px" alt="Intermediate Link Event PropertyPanel" src="images/intermediate_event/intermediate_link_event_property_panel.png">

#### Example

![Intermediate Link Event Example](images/intermediate_event/intermediate_link_event_example.svg)

> In this example a Link Event will be thrown after the `Do some stuff` Task finished which then will trigger the Intermediate Link Catch Event in `Lane B` so that the `Do some more stuff` Task gets executed.

Reviewer: ```suggestion
In this example a Link Event will be thrown after the `Do some stuff` Task finished.
The Process Instance will then continue at the Intermediate Link Catch Event in `Lane B`.

### Intermediate Timer Event

![Intermediate Timer Event](images/intermediate_event/intermediate_timer_event.svg)

> The Intermediate Timer Event is used to wait for a specific amount of time or until a specified date.

Reviewer: ```suggestion
The Intermediate Timer Event is used to pause the Process Instance for a given amount of time, or until a certain date has been reached.

Unlike other Intermediate Events, Timer Events only come with a `Timer Catch Event`, but not with a corresponding Throw Event.
```

> #### Configuration

Reviewer: ```suggestion
#### How to use it
```

> In order to configure a Intermediate Timer Event two information a needed:
> 
> At first the type must be configured, wheter a Date or a Duration should be configured:

Reviewer: ```suggestion
An Intermediate Timer Event is configured with two parameters.

First, the type of timer to execute:
```

<img width="350px" alt="Intermediate Timer Event PropertyPanel Type" src="images/intermediate_event/intermediate_timer_event_property_panel_type.png">

> After that the Timer Definition must be configured, which tells the process engine when the Intermediate Timer Event should finish:

Reviewer: ```suggestion
Second, the value of the timer.
For a `Duration` timer,  it's the amount of time that the Event should wait.
For a `Date` timer, it's the date and time at which the Event should continue.
```

<img width="350px" alt="Intermediate Timer Event PropertyPanel Timer Definition" src="images/intermediate_event/intermediate_timer_event_property_panel_timer_definition.png">

#### Example

![Intermediate Timer Event Example](images/intermediate_event/intermediate_timer_event_example.svg)

> In this example after the `Do some stuff` Task finished a Intermediate Timer Event is triggered which will pause the process for 5 seconds.
> After the 5 seconds the Intermediate Timer Event is finished and the `Do some more stuff` Task is executed.

Reviewer: ```suggestion
In this example, an Intermediate Timer Event is triggered, after the `Do some stuff` task has finished. 
It's a duration timer that causes the Process Instance for 5 seconds.
After the 5 seconds are over, the Process Instance continues with the `Do some more stuff` task.
```

# Intermediate Event

Intermediate Events are a type of event that can occur mid-process.

There are two different kinds of Intermediate Events: Intermediate Catch Events and Intermediate Throw Events.

> ## Intermediate Catch Event
> 
> Intermediate Catch Events are used to wait for a specific event.
> 
> Unlike a Receive Task, an Intermediate Catch Event does not notify anybody that it has received the event.
> It just waits for the event and continues, when it is received. It does not matter who triggered the event.
> 
> ## Intermediate Throw Event
> 
> Intermediate Throw Events are used to trigger events.
> 
> Unlike a Send Task, an Intermediate Throw Event does not wait for an Intermediate Catch Event to acknowledge receiving the event.
> It just sends the event and moves on.
> 
> Note that an Event triggered by an Intermediate Throw Event can also be caught by a Boundary Event of equal type.

Reviewer: ```suggestion
## Intermediate Throw Event

Intermediate Throw Events are used to trigger events.

Unlike a Send Task, an Intermediate Throw Event does not wait for an Intermediate Catch Event to acknowledge receiving the event.
It just sends the event and moves on.

Note that an Event triggered by an Intermediate Throw Event can also be caught by a Boundary Event of equal type.

## Intermediate Catch Event

Intermediate Catch Events are used to wait for a specific event.

Unlike a Receive Task, an Intermediate Catch Event does not notify anybody that it has received the event.
It just waits for the event and continues, when it is received. It does not matter who triggered the event.
```
